### PR TITLE
introduce zstd compression (for data import/export)

### LINKED ISSUE #16791: clickhouse support zstd compression file for file table function ?
Hello,

dosen't clickhouse support read to zstd file ?

Due to the large number of files, it is considered to compress the files into zstd and use Clickhouse to read them.eg: user_logs_20201108000000_190081200_124_1.zst

eg:
```
aaa|bbb|ccc
....
```

i want use sql to read it:
```
select * from file('user_logs/*20201108*.zst','CSV','xxx String,xxx String,xxx String,xxx String,xxx String,xxx String,xxx String ,xxx String,xxx String') limit 5;
```

but clickhouse no parse it 
```
Code: 27. DB::Exception: Received from localhost:9800. DB::Exception: Cannot parse input: expected '|' before: '\n??PmD0Cw?-?M?Xbil??|#6??jg*????b??R?@????\\???ޓ?C???y??\bg??އ?T\n(4?4fs<Y?v????R??9?Q?0xgA?\bA#Q?r\n?@??*?G??+?n鎱8??-?????H(Fe?1?g?N6': (at row 1)
```

### PR DESCRIPTION
**Pull request by fibersel**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
add ```*.zst``` compression/decompression support.It enables using ```*.zst``` in ```file()``` function and ```Content-encoding: zstd``` in http client.This closes [#16791 ](https://github.com/ClickHouse/ClickHouse/issues/16791)

### CODE REVIEW

**Comment by nikitamikhaylov** [comment]
JFYI: You can write `git commit --allow-empty -m 'Bump'`

**Comment by fibersel** [comment]
@alexey-milovidov @akuzm Can someone merge or review this MR?
It seems that fallen test don't belong to my MR.

**Review by alexey-milovidov — COMMENTED** [review]
How we can control compression level when data is compressed in HTTP interface or when file is written?

For ZLib it's controlled by `http_zlib_compression_level` but:
- there are no such settings for xz, br, zst...
- there is no way to control it for files.

We can think of it and make changes in another PR.

File: src/IO/CompressionMethod.h [hunk 1]
```diff
@@ -28,6 +28,9 @@ enum class CompressionMethod
     /// LZMA2-based content compression
     /// This option corresponds to HTTP Content-Encoding: xz
     Xz,
+    /// Zstd compressor
+    ///  This option corresponds to HTTP Content-Encoding: zstd
```
Comment (alexey-milovidov) [inline src/IO/CompressionMethod.h:32]: zst?

File: src/Server/HTTPHandler.cpp [hunk 2]
```diff
@@ -395,32 +380,30 @@ void HTTPHandler::processQuery(
...
 
                 return next_buffer;
             };
 
             cascade_buffer2.emplace_back(push_memory_buffer_and_continue);
         }
 
-        used_output.out_maybe_delayed_and_compressed = std::make_shared<CascadeWriteBuffer>(
-            std::move(cascade_buffer1), std::move(cascade_buffer2));
+        used_output.out_maybe_delayed_and_compressed
```
Comment (alexey-milovidov) [inline src/Server/HTTPHandler.cpp:405]: Formatted code is obviously worse but it's passable.

File: src/IO/ZstdDeflatingWriteBuffer.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+{
+public:
+    ZstdDeflatingWriteBuffer(
+        std::unique_ptr<WriteBuffer> out_,
+        int compression_level,
+        size_t buf_size = DBMS_DEFAULT_BUFFER_SIZE,
+        char * existing_memory = nullptr,
+        size_t alignment = 0);
+
+    /// Flush all pending data and write zlib footer to the underlying buffer.
```
Comment (alexey-milovidov) [inline src/IO/ZstdDeflatingWriteBuffer.h:22]: Wrong copy-pasted comment (zlib).

File: src/IO/ZstdDeflatingWriteBuffer.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+ZstdDeflatingWriteBuffer::ZstdDeflatingWriteBuffer(
+    std::unique_ptr<WriteBuffer> out_, int compression_level, size_t buf_size, char * existing_memory, size_t alignment)
+    : BufferWithOwnMemory<WriteBuffer>(buf_size, existing_memory, alignment), out(std::move(out_))
+{
+    cctx = ZSTD_createCCtx();
+    if (cctx == nullptr)
+        throw Exception(ErrorCodes::ZSTD_ENCODER_FAILED, "zstd stream encoder init failed: zstd version: {}", ZSTD_VERSION_STRING);
+    ZSTD_CCtx_setParameter(cctx, ZSTD_c_compressionLevel, compression_level);
```
Comment (alexey-milovidov) [inline src/IO/ZstdDeflatingWriteBuffer.cpp:17]: These functionc can also return errors.

File: src/IO/ZstdDeflatingWriteBuffer.cpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+}
+
+void ZstdDeflatingWriteBuffer::nextImpl()
+{
+    if (!offset())
+        return;
+
+    bool last_chunk = hasPendingData();
```
Comment (alexey-milovidov) [inline src/IO/ZstdDeflatingWriteBuffer.cpp:44]: Difficult to understand the meaning of this variable.
How it's possible that "if there is more data it is the last chunk"?

Comment (alexey-milovidov) [reply src/IO/ZstdDeflatingWriteBuffer.cpp:44]: Do we have a test with more than 1MB of data?

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
Minor changes needed.

**Comment by alexey-milovidov** [comment]
It is almost ready.

**Comment by troll2016** [comment]
> Changelog category (leave one):
> 
>     * New Feature
> 
> 
> Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
> add `*.zst` compression/decompression support.It enables using `*.zst` in `file()` function and `Content-encoding: zstd` in http client.This closes [#16791 ](https://github.com/ClickHouse/ClickHouse/issues/16791)

that`s great!!!

**Review by alexey-milovidov — APPROVED** [review]

File: tests/queries/0_stateless/01059_storage_file_brotli.sql [hunk 6]
```diff
@@ -23,9 +23,18 @@ SELECT count(), max(x) FROM file;
 
 DROP TABLE file;
 
-SELECT count(), max(x) FROM file('data{1,2,3}.tsv.{gz,br,xz}', TSV, 'x UInt64');
+CREATE TABLE file (x UInt64) ENGINE = File(TSV, 'data4.tsv.zst');
+TRUNCATE TABLE file;
+
+INSERT INTO file SELECT * FROM numbers(1000000);
```
Comment (bharatnc) [inline tests/queries/0_stateless/01059_storage_file_brotli.sql:29]: @evillique I've noticed this behavior (before as well while I was doing the initial implementation) where inserting into file / table with large number of rows causes the query to hang. So I added this test (similar to other compression methods) to confirm and it still hangs according to my local test.

Can also reproduce using (query never completes):

```
tower :) insert into function file('test2.lz4', 'TSV', 'number UInt64', 'lz4') select * from numbers(0,1000000);

INSERT INTO FUNCTION file('test2.lz4', 'TSV', 'number UInt64', 'lz4') SELECT *
FROM numbers(0, 1000000)

Query id: 3fa90fbe-9bab-4d92-9aaf-79763ddc5de1

→ Progress: 1.00 million rows, 8.00 MB (9.12 million rows/s., 72.99 MB/s.)  99%
```

**Pull request merged**